File: packages/node/src/routing/handlers/inline-query.ts
import type { Update } from "../../kernel/update.js";
import { CallbackContext } from "../../kernel/context.js";
import { BaseHandler, HandlerCallback } from "./base.js";

/**
 * Handler for Telegram inline queries triggered when a user types `@bot query` in any chat.
 *
 * @typeParam C - Type of the callback context.
 * @typeParam R - Return value type.
 * @example
 * ```ts
 * const inlineHandler = new InlineQueryHandler(/^search:(.+)$/, async (update, context) => {
 *   const query = update.inline_query!.query;
 *   await context.bot.answerInlineQuery({
 *     inline_query_id: update.inline_query!.id,
 *     results: [],
 *   });
 * });
 * ```
 */
export class InlineQueryHandler<
  C extends CallbackContext = CallbackContext,
  R = unknown,
> extends BaseHandler<C, R> {
  /**
   * Optional pattern or predicate string/regex matching `inline_query.query`.
   */
  public readonly pattern?: RegExp | string | ((query: string) => boolean);

  /**
   * Constructs a new {@link InlineQueryHandler}.
   *
   * @param callbackOrPattern - Callback function or string/RegExp match pattern.
   * @param callbackOrPattern2 - Callback function or string/RegExp match pattern.
   */
  constructor(
    callbackOrPattern:
      HandlerCallback<C, R> | RegExp | string | ((query: string) => boolean) | null | undefined,
    callbackOrPattern2?:
      HandlerCallback<C, R> | RegExp | string | ((query: string) => boolean) | null | undefined,
  ) {
    let cb: HandlerCallback<C, R>;
    let pat: RegExp | string | ((query: string) => boolean) | undefined;
    if (
      typeof callbackOrPattern === "function" &&
      callbackOrPattern.length >= 1 &&
      callbackOrPattern2 === undefined
    ) {
      cb = callbackOrPattern as HandlerCallback<C, R>;
    } else if (
      typeof callbackOrPattern === "function" &&
      typeof callbackOrPattern2 !== "function"
    ) {
      cb = callbackOrPattern as HandlerCallback<C, R>;
      pat = (callbackOrPattern2 as RegExp | string | ((query: string) => boolean)) ?? undefined;
    } else {
      cb = callbackOrPattern2 as HandlerCallback<C, R>;
      pat = (callbackOrPattern as RegExp | string | ((query: string) => boolean)) ?? undefined;
    }

    super(cb);
    this.pattern = pat;
  }

  /**
   * Checks whether the update contains a matching inline query.
   *
   * @param update - The update to test.
   * @returns `true` if inline query matches, `false` otherwise.
   */
  async checkUpdate(update: Update): Promise<boolean> {
    const query = update.inline_query;
    if (!query) return false;
    if (!this.pattern) return true;
    const text = query.query;
    if (typeof this.pattern === "string") {
      return text === this.pattern;
    }

    if (this.pattern instanceof RegExp) {
      return this.pattern.test(text);
    }

    if (typeof this.pattern === "function") {
      return Boolean(this.pattern(text));
    }

    return false;
  }

  /**
   * Populates `context.matches` if using a RegExp pattern and executes callback.
   *
   * @param update - The incoming update.
   * @param context - Callback context instance.
   * @returns Result from callback execution.
   */
  override async handleUpdate(update: Update, context: C): Promise<R> {
    const query = update.inline_query;
    if (query && this.pattern instanceof RegExp) {
      const matches = query.query.match(this.pattern);
      if (matches) {
        context.matches = [matches];
      }
    }

    return super.handleUpdate(update, context);
  }
}

/**
 * Handler for tracking results chosen by users from inline queries.
 *
 * @typeParam C - Type of the callback context.
 * @typeParam R - Return value type.
 * @example
 * ```ts
 * const chosenHandler = new ChosenInlineResultHandler(async (update, context) => {
 *   console.log("Chosen result ID:", update.chosen_inline_result?.result_id);
 * });
 * ```
 */
export class ChosenInlineResultHandler<
  C extends CallbackContext = CallbackContext,
  R = unknown,
> extends BaseHandler<C, R> {
  /**
   * Optional pattern or predicate string/regex matching `result_id` or `query`.
   */
  public readonly pattern?: RegExp | string | ((resultId: string) => boolean);

  /**
   * Constructs a new {@link ChosenInlineResultHandler}.
   *
   * @param callbackOrPattern - Callback function or string/RegExp match pattern.
   * @param callbackOrPattern2 - Callback function or string/RegExp match pattern.
   */
  constructor(
    callbackOrPattern:
      HandlerCallback<C, R> | RegExp | string | ((resultId: string) => boolean) | null | undefined,
    callbackOrPattern2?:
      HandlerCallback<C, R> | RegExp | string | ((resultId: string) => boolean) | null | undefined,
  ) {
    let cb: HandlerCallback<C, R>;
    let pat: RegExp | string | ((resultId: string) => boolean) | undefined;
    if (
      typeof callbackOrPattern === "function" &&
      callbackOrPattern.length >= 1 &&
      callbackOrPattern2 === undefined
    ) {
      cb = callbackOrPattern as HandlerCallback<C, R>;
    } else if (
      typeof callbackOrPattern === "function" &&
      typeof callbackOrPattern2 !== "function"
    ) {
      cb = callbackOrPattern as HandlerCallback<C, R>;
      pat = (callbackOrPattern2 as RegExp | string | ((resultId: string) => boolean)) ?? undefined;
    } else {
      cb = callbackOrPattern2 as HandlerCallback<C, R>;
      pat = (callbackOrPattern as RegExp | string | ((resultId: string) => boolean)) ?? undefined;
    }

    super(cb);
    this.pattern = pat;
  }

  /**
   * Checks whether the chosen inline result matches the pattern.
   *
   * @param update - The update to test.
   * @returns `true` if chosen result matches, `false` otherwise.
   */
  async checkUpdate(update: Update): Promise<boolean> {
    const chosen = update.chosen_inline_result;
    if (!chosen) return false;
    if (!this.pattern) return true;
    const target = chosen.query || chosen.result_id;
    if (typeof this.pattern === "string") {
      return chosen.result_id === this.pattern || chosen.query === this.pattern;
    }

    if (this.pattern instanceof RegExp) {
      return this.pattern.test(target);
    }

    if (typeof this.pattern === "function") {
      return Boolean(this.pattern(target));
    }

    return false;
  }

  /**
   * Populates `context.matches` if using a RegExp pattern and executes callback.
   *
   * @param update - The incoming update.
   * @param context - Callback context instance.
   * @returns Result from callback execution.
   */
  override async handleUpdate(update: Update, context: C): Promise<R> {
    const chosen = update.chosen_inline_result;
    if (chosen && this.pattern instanceof RegExp) {
      const target = chosen.query || chosen.result_id;
      const matches = target.match(this.pattern);
      if (matches) {
        context.matches = [matches];
      }
    }

    return super.handleUpdate(update, context);
  }
}

File: packages/python/src/telebot_py/routing/handlers/inline_query.py
"""Handlers for inline queries and chosen inline results."""

from __future__ import annotations

import re
from collections.abc import Callable
from typing import Any

from telebot_py.routing.handlers.base import BaseHandler, HandlerCallback
from telebot_py.types import Update

#: A pattern tested against the inline query text (or chosen-result target):
#: a string (exact equality), a compiled regular expression (search
#: semantics), or a predicate callable. Mirrors the node pattern option.
InlineQueryPattern = str | re.Pattern[str] | Callable[[str], bool]


class InlineQueryHandler(BaseHandler):
    """Handler for inline queries triggered by typing ``@bot query``.

    With no pattern every inline query matches. Otherwise the query text must
    satisfy the pattern: exact equality for strings, ``search`` semantics for
    compiled regular expressions, a plain call for callable patterns —
    mirroring the node ``InlineQueryHandler``. A regex match is exposed on
    the callback context as ``context.matches``.

    Attributes:
        pattern: The pattern constraining which inline queries match.
    """

    def __init__(
        self,
        pattern: InlineQueryPattern | None,
        callback: HandlerCallback,
        *,
        block: bool = True,
    ) -> None:
        """Initialize the handler.

        Args:
            pattern: Pattern constraining ``inline_query.query``, or ``None``
                to match every inline query.
            callback: Callable invoked when a query matches.
            block: Whether dispatch awaits the callback before moving on.
        """
        super().__init__(callback, block=block)
        self.pattern = pattern

    def check_update(self, update: object) -> bool:
        """Return whether the update carries a matching inline query.

        Args:
            update: The incoming update.

        Returns:
            ``True`` when the inline query satisfies the pattern.
        """
        if not isinstance(update, Update) or update.inline_query is None:
            return False
        if self.pattern is None:
            return True
        text = update.inline_query.query
        if isinstance(self.pattern, str):
            return text == self.pattern
        if isinstance(self.pattern, re.Pattern):
            return self.pattern.search(text) is not None
        return bool(self.pattern(text))

    def collect_additional_context(self, context: Any, update: Update, check_result: Any) -> None:
        """Store the regex match on ``context.matches`` when applicable.

        Args:
            context: The callback context being populated.
            update: The update being processed.
            check_result: The value :meth:`check_update` returned.
        """
        if not check_result or not isinstance(update, Update):
            return
        query = update.inline_query
        if query is None or not isinstance(self.pattern, re.Pattern):
            return
        match = self.pattern.search(query.query)
        if match is not None:
            context.matches = [match]


class ChosenInlineResultHandler(BaseHandler):
    """Handler for results chosen by users from inline queries.

    With no pattern every chosen result matches. Otherwise the target — the
    chosen result's ``query`` when non-empty, else its ``result_id`` — must
    satisfy the pattern: string patterns match either ``result_id`` or
    ``query`` exactly, compiled regular expressions use ``search`` semantics
    on the target, callable patterns receive the target — mirroring the node
    ``ChosenInlineResultHandler``. A regex match is exposed on the callback
    context as ``context.matches``.

    Attributes:
        pattern: The pattern constraining which chosen results match.
    """

    def __init__(
        self,
        pattern: InlineQueryPattern | None,
        callback: HandlerCallback,
        *,
        block: bool = True,
    ) -> None:
        """Initialize the handler.

        Args:
            pattern: Pattern constraining the chosen result's ``result_id``
                or ``query``, or ``None`` to match every chosen result.
            callback: Callable invoked when a chosen result matches.
            block: Whether dispatch awaits the callback before moving on.
        """
        super().__init__(callback, block=block)
        self.pattern = pattern

    def check_update(self, update: object) -> bool:
        """Return whether the update carries a matching chosen result.

        Args:
            update: The incoming update.

        Returns:
            ``True`` when the chosen inline result satisfies the pattern.
        """
        if not isinstance(update, Update) or update.chosen_inline_result is None:
            return False
        chosen = update.chosen_inline_result
        if self.pattern is None:
            return True
        if isinstance(self.pattern, str):
            return chosen.result_id == self.pattern or chosen.query == self.pattern
        target = chosen.query or chosen.result_id
        if isinstance(self.pattern, re.Pattern):
            return self.pattern.search(target) is not None
        return bool(self.pattern(target))

    def collect_additional_context(self, context: Any, update: Update, check_result: Any) -> None:
        """Store the regex match on ``context.matches`` when applicable.

        Args:
            context: The callback context being populated.
            update: The update being processed.
            check_result: The value :meth:`check_update` returned.
        """
        if not check_result or not isinstance(update, Update):
            return
        chosen = update.chosen_inline_result
        if chosen is None or not isinstance(self.pattern, re.Pattern):
            return
        target = chosen.query or chosen.result_id
        match = self.pattern.search(target)
        if match is not None:
            context.matches = [match]
